faunadb-java/src/main/java/com/faunadb/client/query/Expr.java

```diff
@@ -0,0 +1,230 @@
+package com.faunadb.client.query;
+
+import com.fasterxml.jackson.annotation.JsonValue;
+import com.faunadb.client.types.Ref;
+import com.faunadb.client.types.SetRef;
+import com.faunadb.client.types.Value;
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.time.Instant;
+import java.time.LocalDate;
+import java.util.Map;
+
+import static java.lang.String.format;
+
+/**
+ * A query language expression. Constructors for this class are at the {@link Language} class.
+ */
+public abstract class Expr implements Value {
```

Comment:
As it stands right now, `Expr` is a `Value` and knows a few concrete values in order to do escaping. It's a one direction dependency.

If we make `Value` a subtype of `Expr`, we'll create a cyclic dependency between `query` and `types` packages where they both need to know each others' classes. Unless we move `Expr` to `types` package and make `Value` a nested interface but, we'll still depend on concrete subclasses for escaping. I don't know how I feel about this.

Another option would be to break down escaping logic from the `Expr` class itself. That might work out although we may have to do some up / down class casts on the way, which are not verified in compilation time.

In Scala, we use implicit conversions. That avoids type dependencies.

Also, if an expression must have a `value`, all subclasses would also need to satisfy this contract. It may not make sense for `StringV` to have a method `public Value value() {...}` on it, for example.

I'll explore this a little more and see what are the options. I would appreciate if you could flush out any ideias you might have as well.